pkg/scheduler/plugins/deviceshare/plugin.go

```diff
@@ -0,0 +1,298 @@
+	_ framework.ReservePlugin   = &Plugin{}
+	_ framework.PreBindPlugin   = &Plugin{}
+)
+
+type preFilterState struct {
+	skip                    bool
+	allocationResult        apiext.DeviceAllocations
+	convertedDeviceResource corev1.ResourceList
+}
+
+func (s *preFilterState) Clone() framework.StateData {
+	return s
+}
+
+func (g *Plugin) Name() string {
+	return Name
+}
+
+func (g *Plugin) PreFilter(ctx context.Context, cycleState *framework.CycleState, pod *corev1.Pod) *framework.Status {
+	state := &preFilterState{
+		skip:                    true,
+		convertedDeviceResource: make(corev1.ResourceList),
+	}
+
+	podRequest, _ := resource.PodRequestsAndLimits(pod)
+
+	for deviceType := range deviceResourceNames {
+		switch deviceType {
+		case schedulingv1alpha1.GPU:
+			if !hasDeviceResource(podRequest, deviceType) {
+				break
+			}
+			combination, err := validateGPURequest(podRequest)
+			if err != nil {
+				return framework.NewStatus(framework.Error, err.Error())
+			}
+			state.convertedDeviceResource = quotav1.Add(
+				state.convertedDeviceResource,
+				convertGPUResource(podRequest, combination),
+			)
+			state.skip = false
+		case schedulingv1alpha1.RDMA, schedulingv1alpha1.FPGA:
+			if !hasDeviceResource(podRequest, deviceType) {
+				break
+			}
+			if err := validateCommonDeviceRequest(podRequest, deviceType); err != nil {
+				return framework.NewStatus(framework.Error, err.Error())
+			}
+			state.convertedDeviceResource = quotav1.Add(
+				state.convertedDeviceResource,
+				convertCommonDeviceResource(podRequest, deviceType),
+			)
+			state.skip = false
+		default:
+			klog.Warningf("device type %v is not supported yet", deviceType)
+		}
+	}
+
+	cycleState.Write(stateKey, state)
+	return nil
+}
+
+func (g *Plugin) PreFilterExtensions() framework.PreFilterExtensions {
+	return nil
+}
+
+func getPreFilterState(cycleState *framework.CycleState) (*preFilterState, error) {
+	c, err := cycleState.Read(stateKey)
+	if err != nil {
+		// preFilterState doesn't exist, likely PreFilter wasn't invoked.
+		return nil, fmt.Errorf("reading %q from cycleState: %w", stateKey, err)
+	}
+
+	s, ok := c.(*preFilterState)
+	if !ok {
+		return nil, fmt.Errorf("%+v  convert to gpuShare.preFilterState error", c)
+	}
+	return s, nil
+}
+
+func (g *Plugin) Filter(ctx context.Context, cycleState *framework.CycleState, pod *corev1.Pod, nodeInfo *framework.NodeInfo) *framework.Status {
+	state, err := getPreFilterState(cycleState)
+	if err != nil {
+		return framework.AsStatus(err)
+	}
+	if state.skip {
+		return nil
+	}
+
+	if nodeInfo.Node() == nil {
+		return framework.NewStatus(framework.Error, "node not found")
+	}
+
+	nodeDeviceInfo := g.nodeDeviceCache.getNodeDevice(nodeInfo.Node().Name)
+	if nodeDeviceInfo == nil {
+		return framework.NewStatus(framework.UnschedulableAndUnresolvable, ErrMissingDevice)
+	}
+
+	nodeDeviceInfo.lock.Lock()
+	defer nodeDeviceInfo.lock.Unlock()
+
+	podRequest := state.convertedDeviceResource.DeepCopy()
```

Comment:
Do we need a copy? This is supposed to be read-only, right?

Reply:
`tryAllocate` will modify the value in podRequest.